File: Python/lib/kboard.py
import time
'''
kboard static class returns string, integer, float, boolean and menu choices
Use:
import kboard as kb #shortens the class name to kb

Adding a current console row number at the end of each method overwrites errors on the same line
eg. if the console has already got 5 lines of output:
user_name = kb.get_string("Type your name", True, 5) will display messages over the prompt on line 6
for 2 seconds, then re-display the prompt.

The delay can be over-ridden with kb.delay = 3.5 for a longer delay

user_name = kb.get_string("Type your name", True) # returns user input in Title Case

user_age = kb.get_integer("How old are you", 5, 120) # gets an integer between 5 and 120 from the user

user_height = kb.get_float("How tall are you in metres?", 0.5, 2.5) # gets a float between 0.5 and 2.5 from the user

user_likes_python = kb.get_boolean("Do you like Python? (y/n)") # returns True or False from the user

menu_list = ["Brilliant", "Not bad", "Could do better", "Rubbish"]
user_choice = kb.menu("What do think of this utility?", menu_list)
print(f"User thinks this utility is: {menu_list[user_choice]}")

kb.get_string("Press Enter to Quit", False, 0, 20) # Used instead of input("Press Enter to Quit")
at the end of a file to prevent the console closing (does not apply when run from IDE)
'''
blank = "".ljust(79," ")
delay = 2

def process_input(prompt:str, min_value:int, max_value:int, data_type:str, row) -> str:
	''' This function is not called directly from other files. Python does not have a 'Private' keyword'''
	valid_input:bool = False
	while valid_input is False:
		clear_input_field(row)
		user_input:str = input(prompt + "_").strip()		# remove leading and trailing spaces
		output = user_input									# used later if user_input converted to int/float
		if data_type == "string":							# string requested
			if len(user_input) == 0 and min_value > 0:		# enter pressed when min characters > 0
				error_message(row, "noinput", output)
			elif len(user_input) > max_value:
				error_message(row, "string", output)
			else:
				valid_input = True							# string or "" allowed
		else:												# bool, int or float requested
			if len(user_input) == 0:						# enter pressed
				error_message(row, "noinput", output)
			else:
				if data_type == "bool":						# requesting True/False output
					if user_input[0].lower() == "y":
						user_input = True
						valid_input = True
					elif user_input[0].lower() == "n":
						user_input = False
						valid_input = True
					else:
						error_message(row, "bool", output)
				else:										# int or float requested
					try:
						if data_type == "int":
							user_input = int(user_input)
						elif data_type == "float":
							user_input = float(user_input)				
							
						if user_input >= min_value and user_input <= max_value: # number is within limits
							if round(user_input) != user_input and data_type == "int":
								error_message(row, "notint", output)
							else:
								valid_input = True
						else:
							error_message(row, "range", output, min_value, max_value)
					except:
						if data_type == "int":
							error_message(row, "notint", output) # int or float requested, but input not converted
						else:
							error_message(row, "nan", output)
			
	return user_input # True/False or string or int or float returned
			
def get_string(prompt:str, with_title:bool = False, min_value:int = 1, max_value:int = 20, row:int = -1) -> str: # with_title, min_value and max_value can be over-ridden by calling code
	''' Public method: Gets a string from the user, with options for Title Case, length of the string. Set min_value to 0 to allow empty string return '''
	if row >= 0:
		row += 1
	result = process_input(prompt, min_value, max_value, "string", row)
	if with_title:
		result = result.title()

	return result
	
def get_integer(prompt:str, min_value:int = 0, max_value:int = 65536, row:int = -1) -> int: # min_value and max_value can be over-ridden by calling code
	''' Public Method: gets an integer from the user '''
	if row >= 0:
		row += 1	
	return process_input(prompt, min_value, max_value, "int", row)

def get_float(prompt:str, min_value:float = 0.0, max_value:float = 1000000.0, row:int = -1) -> float: # min_value and max_value can be over-ridden by calling code
	''' Public Method: gets a float from the user '''
	if row >= 0:
		row += 1	
	return process_input(prompt, min_value, max_value, "float", row)
	
def get_boolean(prompt:str, row:int = -1) -> bool:
	''' Public Method: gets a boolean (yes/no) type entries from the user '''
	if row >= 0:
		row += 1	
	return process_input(prompt, 1, 3, "bool", row)

def menu(title:str, menu_list:list, row = -1) -> int:
	''' displays a menu using the text in 'title', and a list of menu items (string) '''
	rows = -1
	if row >= 0:
		rows = row + len(menu_list) + 1	
	print(title)
	for i in range(len(menu_list)):
		if i < 9:
			print(f"     {i+1}) {menu_list[i]}")
		else:
			print(f"    {i+1}) {menu_list[i]}") 
			
	return get_integer(f"Type the number of your choice (1 to {len(menu_list)})", 1, len(menu_list), rows) - 1 # -1 to return correct list index

def error_message(row:int, error_type:str, user_input:str, min_value:float = 0, max_value:float = 0) -> None:
	''' display error message. If row number supplied overwrite with delay '''
	
	message = "Just pressing the Enter key or spacebar doesn't work" # default for "noinput"
	
	if error_type == "string":
		message = f"Try entering text between {min_value} and {max_value} characters"
	elif error_type == "bool":
		message = "Only anything starting with y or n is accepted"
	elif error_type == "notint":
		message = f"Try entering a whole number - {user_input} does not cut it"	
	elif error_type == "nan":
		message = f"Try entering a number - {user_input} does not cut it"
	elif error_type == "range":
		message = f"Try a number from {min_value} to {max_value}"
	
		
	if row > 0:
		message = f">>> {message} <<<"
	else:
		message = f"\n{message}..."	
		
	if row >= 0:
		clear_input_field(row)
	print(message)
	if row >= 0:
		time.sleep(delay)
		clear_input_field(row)

def set_cursor_pos(row:int, col:int) -> None:
	''' Sets the cursor position '''
	print(f"\033[{row};{col}H", end = '')
	
def clear_input_field(row:int) -> None:
	if row >= 0:
		set_cursor_pos(row, 0)
		print(blank)
		set_cursor_pos(row, 0)
		
def sleep(s:float) -> None:
	time.sleep(s)
	

File: C#/Kboard.cs
/*
kboard static class returns string, integer, float, boolean and menu choices.
Use:

int row = Kboard.Clear();
string name = Kboard.GetString("What is your name?", true, 1, 10, row);
int age = Kboard.GetInteger("How old are you", 5, 110, row);
double height = Kboard.GetRealNumber("How tall are you?", 0.5, 2.0, row);
bool likesPython = Kboard.GetBoolean("Do you like C#? (y/n)", row);

List<string> options = new List<string> {"Brilliant", "Not bad", "Could do better", "Rubbish"};
int choice =Kboard.Menu("What do think of this utility?", options, row)
*/
using System;
using System.Collections.Generic;
using System.Globalization;
using System.Linq;
using System.Threading;

namespace KBoard
{
    internal class Kboard
    {
        public static int Delay = 2000;
        public static int Clear()
        {
            /// Clears the Console (Allows calls from other classes not involved with UI) ///
            Console.Clear();
            return 0;
        }
        private static void ClearInputField(int row)
        {
            /// use SetCursorPos to delete a line of text ///
            if (row >= 0)
            {
                Console.SetCursorPosition(0, row); // left, top 0 based
                Console.Write("".PadRight(Console.WindowWidth - 1));
                Console.SetCursorPosition(0, row);
            }
        }
        private static void ErrorMessage(int row, string errorType, string userInput,  double minValue = 0, double maxValue = 0)
        {
            /// Display error message to the user for <delay> seconds ///
            if (row < 0) row = 0;
            string message = "Just pressing the Enter key or spacebar doesn't work"; // default for "noinput"
            if (errorType == "string")
                message = $"Try entering text between {minValue} and {maxValue} characters";
            else if (errorType == "bool")
                message = "Only anything starting with y or n is accepted";
            else if (errorType == "nan")
                message = $"Try entering a number - {userInput} does not cut it";
            else if (errorType == "notint")
                message = $"Try entering a whole number - {userInput} does not cut it";
            else if (errorType == "range")
                message = $"Try a number from {minValue} to {maxValue}";
            else if (errorType == "intconversion")
                message = $"Try entering a whole number: {userInput} cannot be converted...";
            else if (errorType == "realconversion")
                message = $"Try entering a decimal number: {userInput} cannot be converted...";

            message = $">>> {message} <<<";
            ClearInputField(row);           // clear current line
            Console.Write(message);         // write message
            Thread.Sleep(Delay);            // pause 2 seconds
            ClearInputField(row);           // clear current line
        }
        public static bool GetBoolean(string prompt, int row = -1)
        {
            /// gets a boolean (yes/no) type entry from the user
            string userInput = ProcessInput(prompt, 1, 3, "bool", row);
            return Convert.ToBoolean(userInput);
        }
        public static double GetRealNumber(string prompt, double min, double max, int row = -1)
        {
            /// gets a float / double from the user
            string userInput = ProcessInput(prompt, min, max, "real", row);
            return Convert.ToDouble(userInput);
        }
        public static int GetInteger(string prompt, double min, double max, int row = -1)
        {
            /// Public Method: gets an integer from the user ///
            string userInput = ProcessInput(prompt, min, max, "int", row);
            return Convert.ToInt32(userInput);
        }
        public static string GetString(string prompt, bool withTitle, int min, int max, int row = -1)
        {
            /// Public Method: gets a string from the user ///
            string userInput = ProcessInput(prompt, min, max, dataType:"string", row);
            if (withTitle)
            {
                TextInfo textInfo = new CultureInfo("en-UK", false).TextInfo;
                userInput = textInfo.ToTitleCase(userInput);
            }
            return userInput;
        }
        public static string Input(string prompt, string ending = "_")
        {
            /// Get keyboard input from user (requires Enter) same as Python ///
            Console.Write($"{prompt}{ending}");
            return Console.ReadLine();
        }
        public static int Menu(string title, List<string> textLines, int row = -1)
        {
            /// displays a menu using the text in 'title', and a list of menu items (string)
            /// This menu will re-draw until user enters correct data
            if (title.Length % 2 == 1) title += " ";
            int rows = -1;
            if (row >= 0) rows = row + textLines.Count + 4;
            int maxLen = title.Length;
            foreach(string line in textLines)
            {
                if(line.Length > maxLen + 9)
                    maxLen = line.Length + 9;
            }
            maxLen += 28;
            if (maxLen > Console.WindowWidth - 2)
                maxLen = Console.WindowWidth - 2;
            if (maxLen % 2 == 1)
                maxLen++;

            string filler = new string(' ', (maxLen - title.Length) / 2);
            Print($"╔{new string('═', maxLen)}╗");
            Print($"║{filler}{title}{filler}║");                     // print title
            Print($"╠{new string('═', maxLen)}╣");
            for (int i = 0; i < textLines.Count; i++)
            {
                if (i < 9)  Print($"║     {i + 1}) {textLines[i].PadRight(maxLen - 8)}║");  // print menu options 5 spaces
                else        Print($"║    {i + 1}) {textLines[i].PadRight(maxLen - 8)}║");   // print menu options 4 spaces
            }
            //Print(new string('═', Console.WindowWidth - 1));
            Print($"╚{new string('═', maxLen)}╝");
            int userInput = GetInteger($"Type the number of your choice (1 to {textLines.Count})",  1, textLines.Count, rows);
            return userInput - 1;
        }
        public static int Print(string text = "")
        {
            /// Replicates Python / Lua print() and returns no of lines printed ///
            Console.WriteLine(text);
            int numLines = text.Count(x => x == '\n') + 1;
            return numLines;
        }
        public static void Sleep(int delay)
        {
            /// replicates Python time.sleep() in seconds ///
            if (delay < 100) delay *= 1000;
            Thread.Sleep(delay);
        }
        private static string ProcessInput(string prompt, double min, double max, string dataType, int row)
        {
            /// validate input, raise error messages until input is valid  ///
            bool valid = false;
            string userInput = "";
            while (!valid)
            {
                ClearInputField(row);
                userInput = Input(prompt);
                if (dataType == "string")
                {
                    if (userInput.Length == 0 && min > 0) ErrorMessage(row, "noinput", userInput);
                    else if (userInput.Length > max) ErrorMessage(row, "string", userInput, min, max);
                    else valid = true;
                }
                else //integer, float, bool
                {
                    if (userInput.Length == 0)                                      // just Enter pressed
                        ErrorMessage(row, "noinput", userInput);
                    else
                    {
                        if (dataType == "bool")
                        {
                            if (userInput.Substring(0, 1).ToLower() == "y")
                            {
                                userInput = "true";
                                valid = true;
                            }
                            else if (userInput.Substring(0, 1).ToLower() == "n")
                            {
                                userInput = "false";
                                valid = true;
                            }
                            else  ErrorMessage(row, "bool", userInput);
                        }
                        if (dataType == "int" || dataType == "real")                // integer / float / double
                        {
                            if (double.TryParse(userInput, out double conversion))  // is a number!
                            {
                                if (conversion >= min && conversion <= max)         // within range!
                                {
                                    if (dataType == "int")                          // check if int
                                    {
                                        if (int.TryParse(userInput, out int intconversion))
                                        {
                                            valid = true;                           // userInput can be converted to int
                                        }
                                        else ErrorMessage(row, "notint", userInput);// real number supplied
                                    }
                                    else valid = true;                              // userInput can be converted to float/double/decimal
                                }
                                else ErrorMessage(row, "range", userInput, min, max);// out of range
                            }
                            else ErrorMessage(row, "nan", userInput);               // not a number
                        }
                    }
                }
            }
            return userInput; //string can be converted to bool, int or float
        }
    }
}
